design-proposals/backend-storage-clone-on-migration/backend-storage-clone-on-migration.md

```diff
@@ -0,0 +1,106 @@
+  - Deprecate RWX FS and require RWX block to enable backend-storage
+- Many storage classes have a minimum PVC size, or a large alignment size. Both can lead to larger storage usage than requested. Having multiple PVCs multiplies that overhead.
+
+We have opened RFEs against libvirt for [TPM](https://issues.redhat.com/browse/RHEL-54325) and [EFI](https://issues.redhat.com/browse/RHEL-54326), in case we decide to pursue to this approach in the future.
+
+### Partitioned block support
+
+[A pull request](https://github.com/kubevirt/kubevirt/pull/10346) (updated [here](https://github.com/kubevirt/kubevirt/pull/12337)) explored a not-safely-migratable implementation which partitions the block storage as ext3.  
+Migratability actually appears functional but should be considered unsafe, since 2 pods (source and target virt-launchers) mount the same ext3 filesystem concurrently.  
+Replacing ext3 with `gfs2` could be something to consider to solve this problem.
+
+## Motivation/Goals
+- Users want to be able to use any storage class for backend-storage
+- Users want all VMs with persistent features to be potentially migratable
+
+## Definition of Users
+Any VM owner that wants to use persistent features, such as TPM and/or EFI
+
+## User Stories
+As a user, I want to be able to seamlessly enable persistent features while preserving the migratability of my VMs.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+- When creating a new VM with persistent features, we use either (first match wins):
+  - The storage class specified in the KubeVirt Custom Resource
+  - The storage class marked as the default for virtualization purposes unless it has a StorageProfile that shows only block is supported
+  - The storage class marked as the default for the cluster
+- All new persistent storage PVCs will be created with the name `persistent-state-for-<vm_name>-<random_string>` and the new annotation `persistent-state-for-<vm_name>=true`
+- When starting an existing VM with persistent features, we will look for any PVC with the annotation, or any PVC with the legacy name, for backwards compatibility
+- The volume mode will be Filesystem
+- The access mode will be RWO, unless a StorageProfile exists for the Storage Class and shows only RWX is supported
+  - Note: the CR field documentation will be adjusted to reflect that RWX is no longer needed
+- When a migration is created:
+  - We create a new empty PVC for the target, with the name `persistent-state-for-<vm_name>-<random_string>` and no annotation
+  - In the migration object status, we store the names of both the source and target PVCs (see API section below)
+  - If a migration succeeds, we set the annotation `persistent-state-for-<vm_name>=true` on the new (target) PVC and delete the old (source) PVC, using the source PVC name from the migration object
+  - If a migration fails, we delete the target backend-storage PVC that was just created, using the target PVC name from the migration object
```

Comment:
Does this hold true also with post-copy migration?

Reply:
Yes, migration mode doesn't make a difference here.

Reply:
Your concern is valid, but it's just a part of the dangers of post-copy. Yes, if a guest is ongoing a post-copy migration and, after the handoff, makes changes to the TPM, then before the end of the migration someone trips on the network cable, the changes made to the TPM will be lost. But that's exactly why 1) post-copy comes with a big warning, 2) bitlocker strongly encourages you to write down your recovery key.
You can also imagine the opposite scenario, if we decided to keep the target PVC and the (post-handoff post-copy) migration failed right during a change to the TPM. The PVC on the target would now contain a corrupt state and switching to it would break the VM (that's what I was referring to above). Same concern applies in RWX btw.
Post-copy migration might break stuff, I don't think there's any way around it...

Reply:
>  if a guest is ongoing a post-copy migration and, after the handoff, makes changes to the TPM, then before the end of the migration someone trips on the network cable, the changes made to the TPM will be lost.

That is not quite correct from the QEMU/libvirt POV. In the event of failed post-copy, both the source and dest QEMU will remain in existence, but both in paused state, since neither has the full state in isolation. 

When the network problem is resolved,  you can call virDomainMigrateStartPostCopy with the VIR_MIGRATE_POSTCOPY_RESUME flag, to recover from the problem and complete the migration, without data loss.